File: frontend/src/services/problemService.js
import api from '../api/api';

const problemService = {
  createProblem: async (problem) => {
    const response = await api.post('/problem/create', problem);
    return response.data;
  },

  getProblem: async (id) => {
    const response = await api.get(`/problem/get`, { params: { id } });
    return response.data;
  },

  updateStatus: async (id, status) => {
    // POST /problem/updateStatus with request params id and status
    const response = await api.post('/problem/updateStatus', null, {
      params: { id, status },
    });
    return response.data;
  },

  updateProblem: async (id, problem) => {
    // POST /problem/update with request param id and body problem
    const response = await api.post('/problem/update', problem, {
      params: { id },
    });
    return response.data;
  },

  getProblemsByStatus: async (status) => {
    const response = await api.get(`/problem/getByStatus`, { params: { status } });
    return response.data;
  },

  getAllProblems: async () => {
    const response = await api.get('/problem/getAll');
    return response.data;
  },

  getProblemsByUserId: async (userId) => {
    const response = await api.get(`/problem/getByUserId/${userId}`);
    return response.data;
  },

  getProblemsByCategory: async (category) => {
    const response = await api.get(`/problem/getByCategory/${category}`);
    return response.data;
  },

  getAvailableProblems: async () => {
    const response = await api.get('/problem/available');
    return response.data;
  },
};

export default problemService;

File: backend/src/main/java/com/example/backend/Service/ProblemService.java
package com.example.backend.Service;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.example.backend.Models.Problem;
import com.example.backend.Models.Status;
import com.example.backend.Models.Worker;
import com.example.backend.Repository.ProblemRepository;
import com.example.backend.Utils.XssUtil;
@Service
public class ProblemService {
    @Autowired
    private ProblemRepository problemRepository;
    @Autowired
    private WorkerService workerService;
    @Autowired
    private EmailService emailService;
    // Business logic for problem management
    public void createProblem(Problem problem) {
        // Sanitize the title and description to prevent XSS attacks
        problem.setTitle(XssUtil.sanitize(problem.getTitle()));
        problem.setDescription(XssUtil.sanitize(problem.getDescription()));
        problemRepository.save(problem);
        List<Worker> workers=workerService.getNearbyWorkers(problem.getLatitude(),problem.getLongitude(),5.0,problem.getCategory());
       for (Worker worker : workers) {
            emailService.sendMail(worker.getEmail(),"New Job Available","A new "+ problem.getCategory()+ " job has been posted near you.");
        }
    }

    public Problem getProblem(String id) {
        // Implementation for getting a problem
        return problemRepository.findById(id).orElse(null);
    }

    public void updateStatus(String id, String status) {
        // Implementation for updating problem status
        Problem problem = problemRepository.findById(id).orElse(null);
        if (problem != null) {
            Status newStatus = Status.valueOf(status.toUpperCase());
            if (newStatus == Status.COMPLETED && problem.getStatus() != Status.COMPLETED) {
                if (problem.getWorkerId() != null) {
                    Worker worker = workerService.getWorkerById(problem.getWorkerId());
                    if (worker != null) {
                        Integer solved = worker.getNoOfProblemsSolved();
                        worker.setNoOfProblemsSolved(solved == null ? 1 : solved + 1);
                        workerService.saveWorker(worker);
                    }
                }
            }
            problem.setStatus(newStatus);
            problemRepository.save(problem);
        }
    }

    public void updateProblem(String id, Problem problem) {
        Problem existingProblem = problemRepository.findById(id).orElse(null);
        if (existingProblem != null && existingProblem.getStatus().equals(Status.PENDING)) {
            existingProblem.setTitle(XssUtil.sanitize(problem.getTitle()));
            existingProblem.setDescription(XssUtil.sanitize(problem.getDescription()));
            existingProblem.setLatitude(problem.getLatitude());
            existingProblem.setLongitude(problem.getLongitude());
            existingProblem.setCategory(problem.getCategory());
            problemRepository.save(existingProblem);
        }
    }
    public List<Problem> getProblemsByStatus(Status status) {
        return problemRepository.findByStatus(status);
}
    public List<Problem> getAllProblems() {
        // Implementation for getting all problems
        return problemRepository.findAll();
    }
    public List<Problem> getProblemsByUserId(String userId) {
        // Implementation for getting problems by user ID
        return problemRepository.findByUserId(userId);
    }
    public List<Problem> getProblemsByCategory(String category) {
        // Implementation for getting problems by category
        return problemRepository.findByCategory(category);
    }
    public List<Problem> getAvailableProblems() {
        // Implementation for getting available problems
        return problemRepository.findByStatus(Status.PENDING);
    }
}
